File: backend/bot/keyboards.py
from typing import Optional
from telegram import InlineKeyboardButton, InlineKeyboardMarkup, ReplyKeyboardMarkup
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from models.admin import Admin, _DEFAULT_KEYBOARDS, _DEFAULT_START_CONTENT
from config import ADMIN_ID


async def get_admin_config(db: AsyncSession) -> Admin:
    result = await db.execute(select(Admin).where(Admin.admin_id == ADMIN_ID))
    admin = result.scalar_one_or_none()
    if not admin:
        admin = Admin(admin_id=ADMIN_ID)
        db.add(admin)
        await db.commit()
        await db.refresh(admin)
    return admin


def build_main_keyboard(admin: Admin) -> ReplyKeyboardMarkup:
    keyboards = admin.keyboards or _DEFAULT_KEYBOARDS
    buttons = [[k["text"]] for k in keyboards]
    return ReplyKeyboardMarkup(buttons, resize_keyboard=True)


def build_start_inline_keyboard(admin: Admin) -> Optional[InlineKeyboardMarkup]:
    buttons = (admin.start_content or {}).get("buttons", [])
    if not buttons:
        return None
    rows = []
    for btn in buttons:
        if btn.get("url"):
            rows.append([InlineKeyboardButton(btn["text"], url=btn["url"])])
        else:
            rows.append([InlineKeyboardButton(btn["text"], callback_data=btn.get("action", "noop"))])
    return InlineKeyboardMarkup(rows)


def build_subscribe_keyboard(channel_id: str) -> InlineKeyboardMarkup:
    if channel_id.startswith("@"):
        channel_link = f"https://t.me/{channel_id[1:]}"
    else:
        channel_link = f"https://t.me/c/{channel_id.replace('-100', '')}"
    return InlineKeyboardMarkup([
        [InlineKeyboardButton("📢 点击订阅频道", url=channel_link)],
        [InlineKeyboardButton("✅ 我已订阅", callback_data="check_subscription")],
    ])

File: backend/src/bot/keyboards.js
const { Markup } = require('telegraf');
const Admin = require('../models/Admin');
const config = require('../config');

/**
 * Get current admin config (creates default if none exists)
 */
async function getAdminConfig() {
  const [admin] = await Admin.findOrCreate({
    where: { adminId: config.ADMIN_ID },
    defaults: { adminId: config.ADMIN_ID },
  });
  return admin;
}

/**
 * Build the main reply keyboard from admin config
 */
async function buildMainKeyboard() {
  const admin = await getAdminConfig();
  const buttons = admin.keyboards.map((k) => [k.text]);
  return Markup.keyboard(buttons).resize();
}

/**
 * Build inline keyboard for /start message
 */
async function buildStartInlineKeyboard() {
  const admin = await getAdminConfig();
  if (!admin.startContent.buttons || admin.startContent.buttons.length === 0) {
    return Markup.inlineKeyboard([]);
  }
  const rows = admin.startContent.buttons.map((btn) => {
    if (btn.url) {
      return Markup.button.url(btn.text, btn.url);
    }
    return Markup.button.callback(btn.text, btn.action || 'noop');
  });
  return Markup.inlineKeyboard(rows);
}

/**
 * Build a t.me link for a given Telegram chat + message.
 * Works for both public channels (@username) and private channels (-100xxx).
 */
function buildChannelMessageUrl(chatId, messageId) {
  if (!chatId || !messageId) return null;
  const str = String(chatId);
  if (str.startsWith('@')) {
    // Public channel: https://t.me/channelname/msgid
    return `https://t.me/${str.slice(1)}/${messageId}`;
  }
  // Private channel: -100xxxxxxxxxx → strip -100 prefix
  const numericId = str.replace(/^-100/, '');
  return `https://t.me/c/${numericId}/${messageId}`;
}

/**
 * Build subscription check keyboard for one or more required chats.
 * @param {string|string[]} requiredChats - single chat ID/username or array
 */
function buildSubscribeKeyboard(requiredChats) {
  const chats = Array.isArray(requiredChats) ? requiredChats : [requiredChats].filter(Boolean);
  const subscribeButtons = chats.map((chatId) => {
    const link = chatId.startsWith('@')
      ? `https://t.me/${chatId.slice(1)}`
      : `https://t.me/c/${String(chatId).replace('-100', '')}`;
    const label = chatId.startsWith('@') ? chatId : `频道 ${chatId}`;
    return [Markup.button.url(`📢 订阅 ${label}`, link)];
  });
  subscribeButtons.push([Markup.button.callback('✅ 我已加入，重新检测', 'check_subscription')]);
  return Markup.inlineKeyboard(subscribeButtons);
}

module.exports = {
  getAdminConfig,
  buildMainKeyboard,
  buildStartInlineKeyboard,
  buildSubscribeKeyboard,
  buildChannelMessageUrl,
};
